### PR TITLE
c-btm inference

### PR DESCRIPTION
**Pull request by NourFahmy**
inference code for c-btm - replicating formulas 2 & 3 from [c-BTM paper](https://arxiv.org/pdf/2303.14177.pdf), and tested locally.

kindly inform if anything else is needed!

link to https://github.com/ontocord/MDEL/issues/40

### CODE REVIEW

**Comment by mrcabbage972** [comment]
Hi @NourFahmy,
The goal is to have a script which the user can call with the arguments:

- The names of the models 
- The input data file path
- The output file path

It would be very helpful if the PR would solve the end to end. It's possible to break this up into a few PR's, if you prefer.

**Comment by NourFahmy** [comment]
Hi @mrcabbage972 - thank you for your feedback!

Would writing the results to an outside file be sufficient for the card?

**Comment by NourFahmy** [comment]
kindly note, still needs to be tested! 
latest commit allows user to call with input and output file path and names of models.

as I understand the sequence of tasks that need to be implemented by for c-btm inference are:
1. calculate the distance between the cluster center of a given domain and the tokenized prompt
2. all the distances are evaluated, and the top k are retained and normalized to sum to 1, and then determined as the most relevant domains given the context
3. the prompt is fed to the most relevant domain experts as per step 2 to generate 1 singular new token
4. the token's probability is weighted by the distance in step 2
5. the next token in the sequence is chosen by choosing the one with highest probability 
6. token is added to the sequence, and the process continues until we reach an end token or a max length of sequence

**Comment by kenhktsui** [comment]
@NourFahmy @mrcabbage972 
I will be finishing the PR #61  by this weekend.
As we do not split the dataset in an unsupervised manner, I trained a few classifiers which gives us the weight of each dataset (and therefore each expert that it is trained on) instead of clustering as in the c-BTM paper.

**Comment by ontocord** [comment]
Where are we on this? @NourFahmy @kenhktsui @mrcabbage972 ?

File: scripts/c-btmInference.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+def load_models(model_names):
+    """
+    takes in list of model names and loads model and corresponding tokenizers
+    returns separate lists of models and tokenizers to be accessed by other functions later
+    """
+    models = []
+    tokenizers = []
+
+    for model_name in model_names:
+        model = AutoModelForCausalLM.from_pretrained(model_name)
```
Comment (NourFahmy) [inline scripts/c-btmInference.py:15]: some issues with loading the models and maintaining HF credentials -- had to load models and tokenizers outside of the function

File: scripts/c-btmInference.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+def generateNextTokenFromExpert(prompt, tokenizer, model):
+  """
+  takes in prompt, tokenizer and model which should be passed from topKFilter
+  as topKFilter determines which domains are most relevant given the context
+  returns the predicted token of the model and its corresponding probability
+  """
+  inputs = tokenizer(prompt)
+  print(inputs['input_ids'])
+  sizeOfInputs = len(inputs['input_ids']) 
+  outputs = model(**inputs, max_new_tokens=1,
```
Comment (NourFahmy) [inline scripts/c-btmInference.py:37]: max_new_tokens not a parameter of geoptx -- how can I limit the number of tokens

Comment (ontocord) [reply scripts/c-btmInference.py:37]: can't you do max_length ?

File: scripts/c-btmInference.py (hunk 1 above)
Comment (ontocord) [reply scripts/c-btmInference.py:15]: Ok. good to know. Strange that you can't load. I made all models public now

**Comment by ontocord** [comment]
I merged it. you can keep adding to it as another PR.

**Pull request merged**